### PR TITLE
Add a spell-checking linter

### PR DESCRIPTION
**Pull request by dannosliwcd**
- Fix spell-checking errors detected by codespell
- Add codespell to a new linter job in our GitHub Actions automation
- Add an initial pre-commit hook configuration that runs codespell

**Why both an action and a pre-commit hook?**
Actions can be used to gate PRs that are candidates to merge into dev, so every PR contribution goes through those checks. Pre-commit hooks do not run for anyone who chooses not to opt in to installing them. However, pre-commit hooks are are still valuable to contributors because they let the contributor see whether each individual commit introduces new issues, instead of only seeing those issues all at once when a PR is pushed.

Draft PR #2670 has some examples of other light-weight checks that might be useful as pre-commit and action linters.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# See https://pre-commit.com for more information
+# See https://pre-commit.com/hooks.html for more hooks
+repos:
+-   repo: https://github.com/codespell-project/codespell
```
Comment (Krebrov001) [inline .pre-commit-config.yaml:4]: I think that this is really cool.

**Review by Krebrov001 — DISMISSED** [review]

File: .codespellrc [hunk 2]
```diff
@@ -0,0 +1,3 @@
+[codespell]
+skip = googletest-release*,autom4te.cache,./service/docs/build,./service/contrib,libtool*,ltmain*,./integration/apps/*/*,./.git,./ChangeLog,*.patch,./test/ProfileTableTest.cpp,configure,config.*,depcomp,*.spec
+ignore-words-list = atleast,clos,fom,warmup,affinitize,affinitized,whis
```
Comment (cmcantalupo) [inline .codespellrc:3]: I guess atleast and whis are okay?  Are these referring to code?

**Review by cmcantalupo — DISMISSED** [review]
LGTM, only question really is about why "atleast" is valid.

File: .codespellrc (hunk 2 above)
Comment (dannosliwcd) [reply .codespellrc:3]: Yeah, we use `atleast` in some test code and `whis` in box-and-whisker plotting code. Both names are provided by external packages so we can't rename them.

**Review by cmcantalupo — APPROVED** [review]

**Pull request merged**